### PR TITLE
Add major compiler version check for building hpx as a binary package

### PR DESCRIPTION
**Pull request by diehlpk**
## Proposed Changes 

If we build HPX with `-DHPX_WITH_BUILD_BINARY_PACKAGE=ON` which is used to build binary packages of HPX, e.g. for the fedora ecosystem, only the major compiler version does need to match and not all versions like before

## Any background context you want to provide?

On most Linux build systems all packages are only rebuild at major compiler changes and not for minor or updates. For each minor compiler update, the Hpx package will throw the error that the exact compiler version is not matching. However, it is not necessary to match the exact version as long as the major version matches.

### CODE REVIEW

File: cmake/templates/HPXMacros.cmake.in [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Copyright (c) 2014 Thomas Heller
-# Copyright (c) 2015 Andreas Sch�fer
+# Copyright (c) 2015 Andreas Sch�fer
```
Comment (hkaiser) [inline cmake/templates/HPXMacros.cmake.in]: Please change this to use `ae` (ASCII characters only)

File: cmake/templates/HPXMacros.cmake.in [hunk 2]
```diff
@@ -11,13 +11,41 @@ include(GNUInstallDirs)
...
 function(hpx_check_compiler_compatibility)
-    if(HPX_IGNORE_COMPILER_COMPATIBILITY OR HPX_WITH_BUILD_BINARY_PACKAGE)
+    if(HPX_IGNORE_COMPILER_COMPATIBILITY)
     return()
-  endif()
-
-  if(NOT (${CMAKE_CXX_COMPILER_ID} STREQUAL ${HPX_CXX_COMPILER_ID}
-     AND ${CMAKE_CXX_COMPILER_VERSION} STREQUAL ${HPX_CXX_COMPILER_VERSION}))
-    set(MESSAGE "Compilers do not match. In order to compile HPX application it is"
+    endif()
```
Comment (hkaiser) [inline cmake/templates/HPXMacros.cmake.in]: Indentation looks to be off here.

File: cmake/templates/HPXMacros.cmake.in [hunk 3]
```diff
@@ -11,13 +11,41 @@ include(GNUInstallDirs)
...
+       
+    if(HPX_WITH_BUILD_BINARY_PACKAGE)
+    
+    string(REGEX MATCH "[0-9]+" HPX_MAJOR_VERSION "${HPX_CXX_COMPILER_VERSION}")
+    string(REGEX MATCH "[0-9]+" CXX_MAJOR_VERSION "${CMAKE_CXX_COMPILER_VERSION}")
```
Comment (hkaiser) [inline cmake/templates/HPXMacros.cmake.in]: Could we go with the same naming scheme as used for the other variables, please (i.e. `HPX_CXX_MAJOR_COMPILER_VERSION` and `CMAKE_CXX_MAJOR_COMPILER_VERSION` or similar)?

File: cmake/templates/HPXMacros.cmake.in [hunk 4]
```diff
@@ -11,13 +11,41 @@ include(GNUInstallDirs)
...
+   
+        if( NOT (${HPX_MAJOR_VERSION} STREQUAL ${CXX_MAJOR_VERSION}) )
+        set(MESSAGE "Compilers do not match. In order to compile HPX application it is"
+        " recommended to use the same major version of the compiler as you did for HPX. "
+        "HPX_CXX_COMPILER=${HPX_CXX_COMPILER}, "
+        "CMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}. "
+        "HPX_CXX_COMPILER_ID=${HPX_CXX_COMPILER_ID}, "
+        "CMAKE_CXX_COMPILER_ID=${CMAKE_CXX_COMPILER_ID}. "
+        "HPX_CXX_MAJOR_VERSION=${HPX_MAJOR_VERSION}, "
+        "CMAKE_CXX_COMPILER_VERSION=${CXX_MAJOR_VERSION}."
```
Comment (hkaiser) [inline cmake/templates/HPXMacros.cmake.in]: This text is inconsistent, please make the generated output understandable.

File: cmake/templates/HPXMacros.cmake.in (hunk 1 above)
Comment (diehlpk) [reply cmake/templates/HPXMacros.cmake.in]: Done

File: cmake/templates/HPXMacros.cmake.in (hunk 4 above)
Comment (diehlpk) [reply cmake/templates/HPXMacros.cmake.in]: Done

File: cmake/templates/HPXMacros.cmake.in (hunk 2 above)
Comment (diehlpk) [reply cmake/templates/HPXMacros.cmake.in]: Done

File: cmake/templates/HPXMacros.cmake.in [hunk 5]
```diff
@@ -11,30 +11,61 @@ include(GNUInstallDirs)
...
+    if(HPX_IGNORE_COMPILER_COMPATIBILITY)
+       return()
+    endif()
+       
+    if(HPX_WITH_BUILD_BINARY_PACKAGE)
+    
+    string(REGEX MATCH "[0-9]+" HPX_MAJOR_VERSION "${HPX_CXX_COMPILER_VERSION}")
+    string(REGEX MATCH "[0-9]+" CXX_MAJOR_VERSION "${CMAKE_CXX_COMPILER_VERSION}")
+   
+        if( NOT (${HPX_MAJOR_VERSION} STREQUAL ${CXX_MAJOR_VERSION}) )
```
Comment (msimberg) [inline cmake/templates/HPXMacros.cmake.in]: Should we not still check the compiler id (name, right?)? Probably a bad idea to mix clang 8 and gcc 8.

Comment (diehlpk) [reply cmake/templates/HPXMacros.cmake.in]: Good catch, I will add it.

File: cmake/templates/HPXMacros.cmake.in [hunk 6]
```diff
@@ -11,30 +11,61 @@ include(GNUInstallDirs)
...
+
+        if (HPX_FIND_REQUIRED)
+            hpx_error("${MESSAGE}")
+        else()
+            hpx_warn("${MESSAGE}")
+            set(${CMAKE_FIND_PACKAGE_NAME}_NOT_FOUND_MESSAGE "${MESSAGE}" PARENT_SCOPE)
+            set(${CMAKE_FIND_PACKAGE_NAME}_FOUND False PARENT_SCOPE)
+        endif()
+  
+        endif()
```
Comment (msimberg) [inline cmake/templates/HPXMacros.cmake.in]: This should be unindented.

File: cmake/templates/HPXMacros.cmake.in [hunk 7]
```diff
@@ -11,30 +11,61 @@ include(GNUInstallDirs)
...
-        )
-    if (HPX_FIND_REQUIRED)
-      hpx_error("${MESSAGE}")
+    if(HPX_IGNORE_COMPILER_COMPATIBILITY)
+       return()
+    endif()
+       
+    if(HPX_WITH_BUILD_BINARY_PACKAGE)
+    
+    string(REGEX MATCH "[0-9]+" HPX_MAJOR_VERSION "${HPX_CXX_COMPILER_VERSION}")
```
Comment (msimberg) [inline cmake/templates/HPXMacros.cmake.in]: @diehlpk could you make another pass and check the indentation, please? This should also be indented.

There's also a parenthesis missing somewhere: https://circleci.com/gh/STEllAR-GROUP/hpx/102216?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link.

File: cmake/templates/HPXMacros.cmake.in (hunk 6 above)
Comment (hkaiser) [reply cmake/templates/HPXMacros.cmake.in]: As said before, I believe this whole file needs indentation corrections. Something is off.

File: cmake/templates/HPXMacros.cmake.in [hunk 8]
```diff
@@ -11,30 +11,60 @@ include(GNUInstallDirs)
...
-    if (HPX_FIND_REQUIRED)
-      hpx_error("${MESSAGE}")
+    if(HPX_IGNORE_COMPILER_COMPATIBILITY)
+       return()
+    endif()
+       
+    if(HPX_WITH_BUILD_BINARY_PACKAGE)
+    
+        string(REGEX MATCH "[0-9]+" HPX_MAJOR_VERSION "${HPX_CXX_COMPILER_VERSION}")
+    	string(REGEX MATCH "[0-9]+" CXX_MAJOR_VERSION "${CMAKE_CXX_COMPILER_VERSION}")
```
Comment (msimberg) [inline cmake/templates/HPXMacros.cmake.in]: There's still something wonky here. The first `string(...` isn't aligned and the second has a tab. Sorry to be so annoying with this. Check that your editor isn't giving you tabs by default. Some of the if-statements further down also have funny alignments.

Comment (msimberg) [reply cmake/templates/HPXMacros.cmake.in]: Should be good now.

**Pull request merged**